### DISCUSSION TITLE
How to pass a numpy float array from Python to C#?

### DISCUSSION THREAD

**Discussion by vadimkantorov** [pythonnet/pythonnet#2226] [Q&A]
I have a method in C# which accepts an argument of type `ReadOnlyMemory<float>`.

How can I call this method from Python (using PythonNet) and pass / marshall a NumPy array (of dtype np.float32)? (need not necessarily a zero-copy way)

**Comment by lostmsu** [comment]
You have to take in a pointer or make a custom codec

**Reply by vadimkantorov** [reply]
For copy-path, should sth like `my_cs_function(System.Array[System.Single](numpy_float32_array.tolist()))`?

There is an implicit operator for conversion of `T[]` to `ReadOnlyMemory<T>`: https://learn.microsoft.com/en-us/dotnet/api/system.readonlymemory-1.op_implicit?view=net-7.0#system-readonlymemory-1-op-implicit(t())-system-readonlymemory((-0))

Would such an implicit conversion operator be called? If not, maybe can use `System.ReadOnlyMemory[System.Single](System.Array[System.Single](numpy_float32_array.tolist()))` (as there is a constructor) ?

**Reply by vadimkantorov** [reply]
`System.ReadOnlyMemory[System.Single](System.Array[System.Single](numpy_float32_array.tolist()))`

So this works (and the implicit conversion operator does not work), but not very ideal from Pythom object memory pressure standpoint (it reallocates a Python array with Python number elements). If I remove the `.tolist()` call, the error is `TypeError: 'numpy.float32' value cannot be converted to System.Single`

I wonder if there is a way to obtain a `ReadOnlyMemory<T>` instance from a pointer and a counter without any memory copies? I found https://stackoverflow.com/a/60711882/445810, this works with a memory copy (which is okay, but not ideal)

**Reply by vadimkantorov** [reply]
I think I'm approaching a zero-copy solution. Would you have an advice of how to call an implicit conversion operator from Python?

E.g. I have a `Memory<System.Single>` object in Python and would like to convert it to `ReadOnlyMemory<System.Single>` prior passing into a C# function. There exists this implicit conversion operator https://learn.microsoft.com/en-us/dotnet/api/system.memory-1.op_implicit?view=net-7.0#system-memory-1-op-implicit(system-memory((-0)))-system-readonlymemory((-0)) , but I don't know how to call it

**Reply by vadimkantorov** [reply]
Okay, figured it out by doing the implicit conversion in C#. Please find the complete example in https://gist.github.com/vadimkantorov/c9ab97aff995ea01ea7f6248ddc733bc#file-csharpfrompython-cs-L23-L41 :)

I think I wouldn't need a Codec impl, but if anyone is looking for this, my example should help!

The zero-copy impl requires unsafe compilation though.

It would also be nice if there existed a natively supported way of converting a NumPy dense tensor to a rectangular C# array, currently this hits `TypeError: 'numpy.float32' value cannot be converted to System.Single` (also asked in https://github.com/pythonnet/pythonnet/discussions/1818)
